src/tracker.rs

```diff
@@ -0,0 +1,300 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    iter::FromIterator,
+    sync::Mutex,
+};
+
+use crate::{DogstatsdClient, TagsProvider};
+
+/// See https://www.datadoghq.com/pricing/ and https://docs.datadoghq.com/account_management/billing/custom_metrics/,
+///
+/// 100 seems like a reasonable place to start warning for now
+pub const DEFAULT_TAG_THRESHOLD: usize = 100;
+
+enum TrackerState {
+    Running {
+        seen: BTreeMap<String, Vec<BTreeSet<String>>>,
+        cardinality_count: usize,
+        actions: Vec<ThresholdAction>,
+    },
+    Done,
+}
+
+pub(crate) struct Tracker {
+    /// Threshold at which to take the user defined action, and stop tracking
+    cardinality_threshold: usize,
+
+    /// Our internal state
+    ///
+    /// This will be `None` if the user does not want to track cardinality.
+    state: Option<Mutex<TrackerState>>,
+}
+
+impl Tracker {
+    fn new(cardinality_threshold: usize, actions: Vec<ThresholdAction>) -> Self {
+        Tracker {
+            cardinality_threshold,
+            state: if !actions.is_empty() && cardinality_threshold != 0 {
+                Some(Mutex::new(TrackerState::Running {
+                    actions,
+                    seen: Default::default(),
+                    cardinality_count: 0,
+                }))
+            } else {
+                None
+            },
+        }
+    }
+
+    /// Track the given tags, and take the user defined action if the cardinality threshold is reached.
+    ///
+    /// Returns the tags as-is, so that it can be used in a `DogstatsdClient` call.
+    pub(crate) fn wrap_and_track<S, T>(&self, dd: &impl DogstatsdClient, metric: &str, tags: T) -> T
```

Comment:
Two questions about this function:
- wraps what exactly? The tags into the rewindable iterator?
- do we need all that magic to reduce binary size and allocations? I feel that it's just premature optimization that increases complexity for a small/negligible gain.

Reply:
> do we need all that magic to reduce binary size and allocations?

No! We actually do not. We actually need all this magic to reduce breaking changes 😰

In a perfect world, we would directly take `&[&str]` or similar as an input for tags, but unfortunately we need to be flexible :(

> I feel that it's just premature optimization that increases complexity for a small/negligible gain.

It is difficult to measure; the gains from such optimisations would only be seen (and fairly measured) in production, as the performance difference from reduced allocations are best visible in environments under real-world stress where a lot of allocations and deallocations are taking place, _especially_ when lots of threads are doing this.

Reply:
> We actually need all this magic to reduce breaking changes 😰

Sorry, but I don't see how it relates to the optimization magic here... for example, can we pass the `&[impl AsRef<str>]` directly to `track` to avoid the extra complexity of the rewindable iterator?

Reply:
Basically, if we pass in a `&[impl AsRef<str>]`, the function might have to generate two copies of itself: one for `&str` (for example), and one for `String`. Sometimes the compiler can "monomorphize" this (flatten these two functions down into one) but this isn't guaranteed. And when it doesn't monomorphize, since the function is basically just copy and pasted with slight changes, binary size inflates.

Reply:
Compilation nerd note: monomorphization is the process of taking a generic fn and implementing it for all of the specific types it is called with, not combining those specific implementations into a single set of instructions in the final binary if they happen to be identical (that's a binary size optimization that happens later, and maybe not at all, at this point differences between how rust and C++ are compiled make me no longer able to reason well about what can happen). The binary size inflation you're worried about is the cost of making a function generic rather than relying on either dynamic dispatch or not having any kind of argument variance at all, and it's cost is generally worth paying in all but extremely performance sensitive applications

I also disagree that it's impossible to measure perf except in production... while it's true that you want to test systems under load, for a library like this, artificial benchmarks get you a good amount of the way there, you just have to be careful to make sure you're really testing what you think you're testing (like by building a binary that does some allocations and frees while also measuring how long calls to `incr` etc take, running it under `perf` and seeing if prima_datadog shows up)

With all that being said, I think the cycling iterator is neat, and am basically pro it's use here. it's some complexity, but it's more or less a slightly strangely shaped ring buffer, which isn't so difficult to reason about, and the tests cover it

Reply:
> And when it doesn't monomorphize, since the function is basically just copy and pasted with slight changes, binary size inflates.

Ok, I got what you meant.
Honestly, binary size is the last thing we should care about in this scenario compared to other (imho more important) things.
We don't lose anything by having a slightly bigger binary in our environment (we are talking about a few hundreds of KBs here, maybe a few MBs in the very worst case?), however to avoid this we have extra code that adds some complexity and that we need to maintain...

> I think the cycling iterator is neat

I agree that it's a nice concept too, but why keeping it if it doesn't give much benefit to us?
I don't think that this is bad code at all, but my minimalist side would like to have less things around if possible...

Reply:
Thought about it a little more: with the current setup, the `Cycle` the rewindable iterator wraps has to run the iterator chain (the `as_ref`, which should be extremely cheap, but is it?) every time (since iterators are lazy), as well as tracking the iterator location (both the outer one we use to yield `None` and the inner one inside of the `Cycle`) etc. Without performance numbers, I'm not convinced all this tracking work is worth it to avoid the occasional instruction cache miss we might suffer from having our `fn`s be generic over `&[impl AsRef<str>]` (in fact, I would find it **highly surprising** if code size had a larger impact on performance than two nested layers of ringbuffer style location tracking). I've removed the rewindable iterator for now, and we can add it back later (in another PR perhaps) if we can nail down a significant performance impact (running cargo bench before and after removing the iter showed no significant impact for me, but also there's significant noise there because `incr` is already so fast that actually measuring it is difficult)